### PR TITLE
Backport all features and fixes from master

### PR DESCRIPTION
**Pull request by joesis**
Resolves https://github.com/getlantern/lantern-internal/issues/3662

After this being merged, I guess we can cut the current `master` as something like `0.9.0` and make this `1.0.0` and start versioning this repo.

I've tested the example on all 3 platforms and all seem to working as expected, with one exception that when running with `-webview` option on Windows, the `lxn/walk` app window doesn't respect the window size specified. We can figure it out when we do need to use webview again, in which case we'll unlikely be using `lxn/walk`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
+* Supported on Windows, macOS, and Linux
+* Menu items can be checked and/or disabled
+* Methods may be called from any Goroutine
+
+## API
 
-## Usage
 ```go
 func main() {
-	// Should be called at the very beginning of main().
```
Comment (joesis) [inline README.md:7]: It's no longer a requirement because the package calls `runtime.LockOSThread` in `init` function now.

`README.md` is largely copied from `master`. Will comment on changes like this one.

File: README.md [hunk 2]
```diff
@@ -22,35 +27,55 @@ func onExit() {
...
+
+### Linux
+
+* Building apps requires the `gtk3`, `libappindicator3` and `libwebkit2gtk-4.0-dev` development headers to be installed. For Debian or Ubuntu, you can may install these using:
+
+```sh
+sudo apt-get install libgtk-3-dev libappindicator3-dev libwebkit2gtk-4.0-dev
+```
+
+To build the example, run `go build example/main.go`. Before building, remove `example/rsrc.syso` which is only used on Windows.
```
Comment (joesis) [inline README.md:71]: Or gcc will complain about mixed architecture when building 64 bit targets.

File: example/main.go [hunk 3]
```diff
@@ -10,19 +11,27 @@ import (
 	"github.com/skratchdot/open-golang/open"
 )
 
+var (
+	webview = flag.Bool("webview", false, "show a webview window along with the systray")
```
Comment (joesis) [inline example/main.go:15]: It's to demonstrate that systray can co-exist with 3rd party UI toolkits.

File: systray.go [hunk 4]
```diff
@@ -16,11 +13,22 @@ import (
 )
 
 var (
-	hasStarted = int64(0)
-	hasQuit    = int64(0)
```
Comment (joesis) [inline systray.go:20]: For simplicity I removed `hasStarted` check (the caller is not expected to call `systray.Quit` if systray is not used) and switched the `hasQuit` check with `sync.Once`.

File: systray.go [hunk 5]
```diff
@@ -36,27 +44,41 @@ type MenuItem struct {
...
-	runtime.LockOSThread()
-	atomic.StoreInt64(&hasStarted, 1)
+	Register(onReady, onExit)
+	nativeLoop()
+}
 
+// Register initializes GUI and registers the callbacks but relies on the
+// caller to run the event loop somewhere else. It's useful if the program
+// needs to show other UI elements, for example, webview.
+func Register(onReady func(), onExit func()) {
```
Comment (joesis) [inline systray.go:81]: This is newly added to support co-existing with other toolkits.

Comment (joesis) [reply systray.go:81]: Other changes to this file are copied from `master`.

File: systray_darwin.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package systray
```
Comment (joesis) [inline systray_darwin.go:1]: Same as in `master`

File: systray_darwin.m [hunk 7]
```diff
@@ -17,25 +17,29 @@ @interface MenuItem : NSObject
 {
```
Comment (joesis) [inline systray_darwin.m:17]: Same as in `master`

File: systray_darwin.m [hunk 8]
```diff
@@ -182,9 +209,12 @@ - (void) quit
 
 @end
 
-int nativeLoop(void) {
+void registerSystray(void) {
```
Comment (joesis) [inline systray_darwin.m:212]: Break the logic into two pieces to co-exist with UI toolkits.

File: systray_windows.go [hunk 9]
```diff
@@ -236,18 +236,26 @@ var wt winTray
...
 		WM_LBUTTONUP  = 0x0202
+		WM_COMMAND    = 0x0111
+		WM_ENDSESSION = 0x0016
+		WM_CLOSE      = 0x0010
+		WM_DESTROY    = 0x0002
+		WM_CREATE     = 0x0001
 	)
 	switch message {
+	case WM_CREATE:
+		systrayReady()
```
Comment (joesis) [inline systray_windows.go:249]: Calls `onReady` only when seeing some Windows message which seems more accurate to me

File: systray_windows.go [hunk 10]
```diff
@@ -236,18 +236,26 @@ var wt winTray
...
 	case WM_COMMAND:
-		menuId := int32(wParam)
-		if menuId != -1 {
-			systrayMenuItemSelected(menuId)
+		menuItemId := int32(wParam)
+		// https://docs.microsoft.com/en-us/windows/win32/menurc/wm-command#menus
+		if menuItemId != 0 {
```
Comment (joesis) [inline systray_windows.go:253]: It's a drive-by fix which may solve the root cause of https://github.com/getlantern/lantern-internal/issues/3624

File: systray_windows.go [hunk 11]
```diff
@@ -236,18 +236,26 @@ var wt winTray
...
+			systrayMenuItemSelected(menuItemId)
 		}
+	case WM_CLOSE:
+		pDestroyWindow.Call(uintptr(t.window))
+		t.wcex.unregister()
```
Comment (joesis) [inline systray_windows.go:258]: The current code calls it in `defer` block but the code doesn't block anymore so I figure this is a good place.

File: systray_windows.go [hunk 12]
```diff
@@ -454,45 +492,61 @@ func (t *winTray) addOrUpdateMenuItem(menuId int32, title string, disabled, chec
...
-	res, _, err := pSetMenuItemInfo.Call(
-		uintptr(t.menu),
-		uintptr(menuId),
-		0,
-		uintptr(unsafe.Pointer(&mi)),
-	)
+	var res uintptr
+	menu, exists := t.menus[parentId]
+	if !exists {
+		menu, err = t.convertToSubMenu(parentId)
```
Comment (joesis) [inline systray_windows.go:515]: This happens when adding a submenu for the first time.

File: systray_windows.go [hunk 13]
```diff
@@ -570,31 +626,96 @@ func (t *winTray) showMenu() error {
...
+		if res == 0 {
+			return 0, err
+		}
+		h = windows.Handle(res)
+		t.loadedImages[src] = h
+	}
+	return h, nil
+}
+
+func (t *winTray) iconToBitmap(hIcon windows.Handle) (windows.Handle, error) {
```
Comment (joesis) [inline systray_windows.go:689]: Have to do this because menu icon only supports bitmap.

**Comment by oxtoacart** [comment]
When I run with the webview on Windows, the "Open UI" link doesn't open the UI in the webview but in the system browser.

**Comment by oxtoacart** [comment]
@joesis Why revert the changes for non-windows platforms? I thought our only problem was with lxn/walk?

**Comment by oxtoacart** [comment]
What if you just keep master mostly as-is but replace the Windows portion with the old code plus your changes to support the new API @joesis?

**Comment by joesis** [comment]
The thinking behind removing webview related stuff is that they are different things. Only a few of the users of `systray`, including Lantern at this moment, would use webview. Those who do use can integrate with dedicated tools like https://github.com/zserge/webview given now `systray` has the ability to co-exist with them.

**Comment by joesis** [comment]
> When I run with the webview on Windows, the "Open UI" link doesn't open the UI in the webview but in the system browser.

Sorry I forgot to mention, the example now has a `-webview` option to open a webview window on startup. If you like, I can remove the option and let "open UI" to show the window, as what the `master` currently doing.

**Comment by oxtoacart** [comment]
> The thinking behind removing webview related stuff is that they are different things. Only a few of the users of systray, including Lantern at this moment, would use webview. Those who do use can integrate with dedicated tools like https://github.com/zserge/webview given now systray has the ability to co-exist with them.

In that case, maybe it's better to remove the webview stuff completely?

**Comment by joesis** [comment]
Yeah systray doesn't have any webview stuff. The option in the example is to demo the ability to co-exist with webview or other UI toolkits.

**Comment by joesis** [comment]
I guess I'll revise the example to make it more apparent, like having both "Open in default browser" and "Open in webview window"

**Comment by joesis** [comment]
Oh well thinking again I'd prefer leave the code as it is more clean and informational by using option.

**Comment by oxtoacart** [comment]
> Yeah systray doesn't have any webview stuff

Ah sorry, I didn't notice that was just in the example :)

**Comment by oxtoacart** [comment]
> Those who do use can integrate with dedicated tools like https://github.com/zserge/webview given now systray has the ability to co-exist with them.

This doesn't seem to work for me (see #143)

**Comment by joesis** [comment]
> > Those who do use can integrate with dedicated tools like https://github.com/zserge/webview given now systray has the ability to co-exist with them.
> 
> This doesn't seem to work for me (see #143)

Yeah it crashes to me as well. Do we need to use webview very soon or can we leave it for investigation in the future? Or you'd rather want to be more confident that systray can co-exist with at least one of such tools on all 3 platforms?

If you prefer, I can also extract the webview related code in the current master which is proved working to a separate repo and update the example to use it.

**Comment by oxtoacart** [comment]
@joesis If I understand the goal of your refactoring, its purpose is to make systray compatible with other frameworks that have their own run loop (like zserge/webview). If that's the case, I think we should make sure that it actually accomplishes this, even if we don't immediately need webview. Otherwise, we could just start with what's on master and keep the current MacOS and Linux code, and bring back the original Windows code plus your support for the new sub-menu stuff.

**Comment by joesis** [comment]
Okay cool let me improve the example (maybe extract to a separate one) to include webview on all 3 platforms, possibly reusing the working code in `master`.

**Comment by joesis** [comment]
@oxtoacart I extracted a separate `webview_example` which basically copied `systray_browser_*` files, and verified on both macOS and Windows 7 VM. On Ubuntu 20 VM I got SIGABRT inside the libwebkit2gtk so. As the code in `master` crashes the same way I didn't dig deeper into it.

**Comment by oxtoacart** [comment]
>  extracted a separate webview_example which basically copied systray_browser_* files, and verified on both macOS and Windows 7 VM. On Ubuntu 20 VM I got SIGABRT inside the libwebkit2gtk so. As the code in master crashes the same way I didn't dig deeper into it.

@joesis Cool!

**Comment by joesis** [comment]
Thanks @oxtoacart !!! Will tag this as 1.0.0 and update README accordingly.
